### PR TITLE
[ModelingOutput] add tinybert/Electra/XLNet/ALBERT/ERNIE-M more output & loss

### PR DESCRIPTION
**Pull request by wj-Mcat**
### PR types
New features

### PR changes
Models

### Description

add more output and label/loss for tinybert model

### CODE REVIEW

File: paddlenlp/transformers/albert/modeling.py [hunk 1]
```diff
@@ -1076,23 +1132,24 @@ def forward(
...
-                "last_hidden_state": sequence_output,
-                "pooler_output": pooled_output,
-                "all_hidden_states": encoder_outputs["all_hidden_states"],
-                "all_attentions": encoder_outputs["all_attentions"],
-            }
-        return sequence_output, pooled_output
+        if not return_dict:
+            return (sequence_output, pooled_output) + encoder_outputs[1:]
+
+        return BaseModelOutputWithPoolingAndCrossAttentions(
```
Comment (guoshengCS) [inline paddlenlp/transformers/albert/modeling.py]: 再check下呢，我看HF这里是`BaseModelOutputWithPooling`

Comment (wj-Mcat) [reply paddlenlp/transformers/albert/modeling.py]: 在`model_outputs`里面确实是缺少了`BaseModelOutputWithPooling`的实现，这个我稍后都调整一下。

Comment (wj-Mcat) [reply paddlenlp/transformers/albert/modeling.py]: @guoshengCS 我这边都添加了`BaseModelOutput`和`BaseModelOutputWithPooling `然后在也调整了对应的模型。

File: paddlenlp/transformers/albert/modeling.py [hunk 2]
```diff
@@ -316,32 +361,32 @@ def __init__(
...
-        head_mask=None,
-        return_dict=False,
-    ):
-        layer_hidden_states = ()
-        layer_attentions = ()
+    def forward(self,
+                hidden_states,
+                attention_mask=None,
+                head_mask=None,
+                output_attentions=False):
```
Comment (guoshengCS) [inline paddlenlp/transformers/albert/modeling.py]: 这里没有output_hidden_states吗，看HF的AlbertLayerGroup也是支持output_hidden_states的

File: paddlenlp/transformers/albert/modeling.py [hunk 3]
```diff
@@ -380,19 +425,22 @@ def __init__(
...
+                head_mask=None,
+                output_hidden_states=False,
+                output_attentions=False,
+                return_dict=False):
         hidden_states = self.embedding_hidden_mapping_in(hidden_states)
 
-        all_hidden_states = (hidden_states, ) if return_dict else None
-        all_attentions = () if return_dict else None
+        all_hidden_states = () if output_hidden_states else None
+        all_attentions = () if output_attentions else None
```
Comment (guoshengCS) [inline paddlenlp/transformers/albert/modeling.py:447]: 建议这种没有使用`paddle.nn.Transformer`的就尽量保持和HF一致吧，看原来这里的代码是和HF一样的

File: paddlenlp/transformers/albert/modeling.py [hunk 4]
```diff
@@ -1368,21 +1451,36 @@ def forward(
             position_ids=position_ids,
             head_mask=head_mask,
             inputs_embeds=inputs_embeds,
-            return_dict=return_dict,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict)
+
+        if paddle.is_tensor(transformer_outputs):
```
Comment (guoshengCS) [inline paddlenlp/transformers/albert/modeling.py]: 这里是否使用`isinstance(transformer_outputs, type(input_ids))`这种方式，以避免潜在的支持动转静和静态图存在的问题，is_tensor只支持动态图使用，其他is_tensor使用也是类似

File: paddlenlp/transformers/electra/modeling.py [hunk 5]
```diff
@@ -958,12 +1043,44 @@ def forward(self,
...
+        sequence_output = self.electra(
+            input_ids,
+            token_type_ids,
+            position_ids,
+            attention_mask,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict)
+
+        if paddle.is_tensor(sequence_output):
```
Comment (guoshengCS) [inline paddlenlp/transformers/electra/modeling.py]: 同上注意对is_tensor的使用

File: paddlenlp/transformers/electra/modeling.py [hunk 6]
```diff
@@ -127,34 +134,50 @@ def forward(self,
...
                 all_attentions.append(output[1])
                 output = output[0]
-            if output_hidden_states:
-                all_hidden_states.append(output)
+
+        if output_hidden_states:
+            all_hidden_states.append(output)
 
         if self.norm is not None:
             output = self.norm(output)
```
Comment (guoshengCS) [inline paddlenlp/transformers/electra/modeling.py:168]: 如果有self.norm的话是不是这样这里也参考 https://github.com/PaddlePaddle/PaddleNLP/blob/develop/paddlenlp/transformers/model_outputs.py#L174 把all_hidden_states中的最后一个换掉，保证和HF语义对齐

File: paddlenlp/transformers/electra/modeling.py [hunk 7]
```diff
@@ -739,20 +783,46 @@ def forward(self,
...
+        generator_sequence_output = self.electra(
+            input_ids,
+            token_type_ids,
+            position_ids,
+            attention_mask,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict)
+
+        if paddle.is_tensor(generator_sequence_output):
```
Comment (guoshengCS) [inline paddlenlp/transformers/electra/modeling.py]: 同上注意is_tensor使用

File: paddlenlp/transformers/electra/modeling.py [hunk 8]
```diff
@@ -1029,13 +1163,36 @@ def forward(self,
...
+            token_type_ids,
+            position_ids,
+            attention_mask,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict)
 
-        sequence_output = self.dropout(sequence_output)
-        logits = self.classifier(sequence_output)
+        if paddle.is_tensor(sequence_output):
```
Comment (guoshengCS) [inline paddlenlp/transformers/electra/modeling.py]: 同上注意对is_tensor的使用

File: paddlenlp/transformers/electra/modeling.py [hunk 9]
```diff
@@ -1547,15 +1723,40 @@ def forward(self,
...
+            input_ids,
+            token_type_ids,
+            position_ids,
+            attention_mask,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+        )
+
+        if paddle.is_tensor(sequence_output):
```
Comment (guoshengCS) [inline paddlenlp/transformers/electra/modeling.py]: 同上注意对is_tensor的使用

File: paddlenlp/transformers/ernie_gram/modeling.py [hunk 10]
```diff
@@ -317,10 +335,28 @@ def forward(self,
...
-        encoder_outputs = self.encoder(embedding_output, attention_mask)
-        sequence_output = encoder_outputs
+        encoder_outputs = self.encoder(
+            embedding_output,
+            attention_mask,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict)
+
+        if paddle.is_tensor(encoder_outputs):
```
Comment (guoshengCS) [inline paddlenlp/transformers/ernie_gram/modeling.py]: 同上注意is_tensor的使用

File: paddlenlp/transformers/tinybert/modeling.py [hunk 11]
```diff
@@ -283,7 +290,13 @@ def set_input_embeddings(self, embedding: nn.Embedding) -> None:
         """
         self.embeddings.word_embeddings = embedding
 
-    def forward(self, input_ids, token_type_ids=None, attention_mask=None):
+    def forward(self,
+                input_ids,
+                token_type_ids=None,
```
Comment (guoshengCS) [inline paddlenlp/transformers/tinybert/modeling.py:293]: 是否也加上position_ids的支持呢

File: paddlenlp/transformers/albert/modeling.py (hunk 2 above)
Comment (wj-Mcat) [reply paddlenlp/transformers/albert/modeling.py]: 嗯嗯，确实是有，我稍后添加上去

File: paddlenlp/transformers/albert/modeling.py (hunk 3 above)
Comment (wj-Mcat) [reply paddlenlp/transformers/albert/modeling.py:447]: 其实是咱们paddlenlp的代码中都是用改后的写法，也只是这个模型的写法不一致，最后的作用都是一样的。

那我还是跟huggingface保持一致吧，稍后我调整过来。

File: paddlenlp/transformers/albert/modeling.py (hunk 4 above)
Comment (wj-Mcat) [reply paddlenlp/transformers/albert/modeling.py]: OK，这部分我调整一下

File: paddlenlp/transformers/electra/modeling.py (hunk 5 above)
Comment (wj-Mcat) [reply paddlenlp/transformers/electra/modeling.py]: Done

File: paddlenlp/transformers/electra/modeling.py (hunk 6 above)
Comment (wj-Mcat) [reply paddlenlp/transformers/electra/modeling.py:168]: 嗯嗯，这部分我调整了。

File: paddlenlp/transformers/electra/modeling.py (hunk 7 above)
Comment (wj-Mcat) [reply paddlenlp/transformers/electra/modeling.py]: Done

File: paddlenlp/transformers/tinybert/modeling.py (hunk 11 above)
Comment (wj-Mcat) [reply paddlenlp/transformers/tinybert/modeling.py:293]: 经过和jiaqi沟通之后确定这里是需要加上的

**Comment by wj-Mcat** [comment]
你四天前给出的comment我这边都已经fix了，有时间review一下。 @guoshengCS

**Review by guoshengCS — APPROVED** [review]
LGTM
